Sources/StreamChat/WebSocketClient/Events/MemberEvents.swift

```diff
@@ -4,41 +4,112 @@
 
 import Foundation
 
-public struct MemberAddedEvent: MemberEvent, ChannelSpecificEvent, EventWithPayload {
-    public let memberUserId: UserId
+public struct MemberAddedEvent: MemberEvent, HasCid, HasUser, HasMember {
     public let cid: ChannelId
+    public let user: ChatUser
+    public let member: ChatChannelMember
+    public let createdAt: Date
+}
+
+extension MemberAddedEvent {
+    struct DTO: EventWithPayload {
+        let cid: ChannelId
+        let user: UserPayload
+        let member: MemberPayload
+        let createdAt: Date
+        let payload: Any
+        
+        init(from response: EventPayload) throws {
+            cid = try response.value(at: \.cid)
+            user = try response.value(at: \.user)
+            member = try response.value(at: \.memberContainer?.member)
+            createdAt = try response.value(at: \.createdAt)
+            payload = response
+        }
+    }
     
-    let payload: Any
-    
-    init(from response: EventPayload) throws {
-        memberUserId = try response.value(at: \.memberContainer?.member?.user.id)
-        cid = try response.value(at: \.cid)
-        payload = response
+    init?(event: DTO, session: DatabaseSession) {
+        guard
+            let user = session.user(id: event.user.id),
+            let member = session.member(userId: event.member.user.id, cid: event.cid)
+        else { return nil }
+        
+        self.init(
+            cid: event.cid,
+            user: user.asModel(),
+            member: member.asModel(),
+            createdAt: event.createdAt
+        )
     }
 }
 
-public struct MemberUpdatedEvent: MemberEvent, ChannelSpecificEvent, EventWithPayload {
-    public let memberUserId: UserId
+public struct MemberUpdatedEvent: MemberEvent, HasCid, HasUser, HasMember {
     public let cid: ChannelId
+    public let user: ChatUser
+    public let member: ChatChannelMember
+    public let createdAt: Date
+}
```

Comment:
Another question is.. do we really need these protocols? Why just simply create `MemberUpdatedEvent` with user, member and cid, without the protocols? It looks to me that the protocols are not needed at all.

Reply:
The reason of using `Has***` is mainly to stay aligned with [Android](https://github.com/GetStream/stream-chat-android/blob/880cfccf9dec7f1f088a1380e888cd6dee5a7536/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/events/ChatEvent.kt). 

The use-case on customer side I was thinking of is catching all the events related to a specific model:
```swift
extension DemoChannelListVC: EventsControllerDelegate {
    func eventsController(_ controller: EventsController, didReceiveEvent event: Event) {
        guard let userEvent = event as? HasUser else { return }
        
        let user = userEvent.user
        ...
    }
}
```

Without a bounding protocol they would have to check for concrete event types which not that nice 🤔 On the other hand, we indeed can remove these protocols/make them `internal` to reduce public API changes so there're pros in both approaches.

@nuno-vieira @b-onc @adamrushy @tbarbugli WDYT guys?

Reply:
What if we could use here something like @dynamicMemberLookup? 🤔 

Since an event can have multiple models, and it is pretty dynamic, it sounds like a good use case for @dynamicMemberLookup, what do you guys think? 

The idea is that an Event can have multiple models, so we would store the models in a dictionary or something, and then the customer could access it with `event.user` or `event.channel` it was like you were accessing a dictionary

Reply:
Discussed internally, using `dynamicMemberLookup` would require breaking changes so will keep that option in mind for the next major.